### ISSUE TITLE
fix(bindings-py): to_sarif rejects None entries from analyze() with confusing TypeError

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#341]
> **Resolution (pending merge, branch `fix/issue-341`, commit 70718733):**
> `to_sarif`'s iterable arm now skips `None` entries silently, mirroring
> the existing `AnalysisError` skip. The Rust pyfunction docstring and
> the `.pyi` stub (type signature + prose) call out the new tolerance.
> Two regression tests (mixed `[dict, None]` and pure `[None, None]`)
> pin the contract; both fail against the pre-fix code with the exact
> `TypeError` from this issue, and pass against the fix. See the
> comment thread below for full details.

## Summary

`bca.to_sarif()` raises `TypeError` when its input iterable contains `None`
(the documented return of `bca.analyze()` for generated files), and the
docstring does not warn callers to filter `None` first. This breaks the
natural pattern `bca.to_sarif([bca.analyze(p) for p in paths])` for any
input set that contains a `@generated` / `DO NOT EDIT` / `GENERATED CODE`
file, despite `skip_generated=True` being the documented default on
`analyze`.

## Location

- `big-code-analysis-py/src/sarif.rs:518-535` (`to_sarif` iterable arm)
- `big-code-analysis-py/python/big_code_analysis/_native.pyi:411-472`
  (`to_sarif` docstring)
- `big-code-analysis-py/tests/test_sarif.py` (no test for `None` input)

## Evidence

The Rust iterable path in `to_sarif`:

```rust
for item in result.try_iter()? {
    let item = item?;
    if item.is_instance_of::<PyAnalysisError>() {
        continue;
    }
    let Ok(dict) = item.clone().cast_into::<PyDict>() else {
        // ... return TypeError "expected a result dict or AnalysisError" ...
    };
    collect_offenders(&dict, &thresholds, &mut offenders)?;
}
```

skips `AnalysisError` entries silently but rejects everything else
(including `None`) with a `TypeError`. The docstring on `to_sarif`
states the iterable "yields such dicts and/or :class:`AnalysisError`
instances (the natural shape of :func:`analyze_batch`'s return value)"
— but `bca.analyze()` returns `dict | None`, so the natural shape of
a per-file list comprehension over `analyze()` includes `None` for any
generated file.

Empirical reproduction (from a fresh tree):

```python
import big_code_analysis as bca
import tempfile, os
with tempfile.TemporaryDirectory() as td:
    good = os.path.join(td, "good.py")
    gen  = os.path.join(td, "gen.py")
    open(good, "w").write("def f(): pass\n")
    open(gen,  "w").write("# @generated\ndef g(): pass\n")
    results = [bca.analyze(good), bca.analyze(gen)]
    # results = [{...}, None]
    bca.to_sarif(results, thresholds={"cyclomatic": 0})
    # TypeError: to_sarif expected a result dict or AnalysisError, got NoneType
```

`bca.analyze_batch()` does not exhibit this because it hard-codes
`skip_generated=False` and substitutes `AnalysisError` for any failure.
The asymmetry is between `analyze` + `to_sarif` (the documented per-file
parity path) and `analyze_batch` + `to_sarif` (the never-raise pipeline
path), and the per-file path is the one a user reading the README's
"Usage" snippet would naturally compose.

## Expected Behavior

Either:

1. `to_sarif` accepts and silently skips `None` entries (mirrors the
   `AnalysisError` skip semantics — both represent "no record emitted
   for this file"); OR
2. The `to_sarif` docstring + `_native.pyi` explicitly call out that
   `None` (from `skip_generated=True`) must be filtered before passing,
   and the error message names the fix.

Option 1 is the more user-friendly choice and consistent with the
`AnalysisError` skip-silently behaviour already documented in the same
function.

## Actual Behavior

`to_sarif` raises `TypeError: to_sarif expected a result dict or
AnalysisError, got NoneType`. The message does not mention `analyze`,
`skip_generated`, or how to fix the input — a user has to read the
source to understand why a list returned by their own per-file
comprehension is being rejected.

## Impact

User-facing pitfall in the documented integration pattern. Affects
anyone constructing the `to_sarif` input from per-file `analyze` calls
with default kwargs (which is the README's documented usage). The
failure mode is loud (`TypeError`) rather than silent, but the error
message does not guide the user to the fix.

Test coverage gap: `tests/test_sarif.py` has 30+ tests covering
threshold validation, MappingProxyType rejection, bool rejection, etc.,
but no test for `None` in the input iterable. A future refactor that
changes the `None` handling (either way) has no regression test to
trip.

**Comment by dekobon** [comment]
## General Plan for Fix

### Root Cause

`to_sarif`'s iterable path was written against `analyze_batch`'s
return shape (`list[dict | AnalysisError]`), but the docstring
advertises the function as accepting "any iterable yielding such
dicts and/or :class:`AnalysisError` instances", which a user
reading the per-file `analyze` API will reasonably extend to
include the `None` returns that `analyze()` produces for generated
files with the documented default `skip_generated=True`.

The asymmetry is structural: `analyze_batch` filters `None` out at
the bridge layer (it hard-codes `skip_generated=false`), so its
output never carries `None`. But every per-file path that goes
through `analyze` and uses default kwargs will produce `None`
entries for generated files. The natural composition with
`to_sarif` then breaks.

### Implementation Steps

Recommended (option 1 in the issue body):

1. In `big-code-analysis-py/src/sarif.rs::to_sarif`, the iterable
   loop already skips `PyAnalysisError` via `continue`. Add a
   parallel skip for `None`:

   ```rust
   for item in result.try_iter()? {
       let item = item?;
       if item.is_none() || item.is_instance_of::<PyAnalysisError>() {
           continue;
       }
       // ... existing dict-cast and error path ...
   }
   ```

2. Update the docstring on `to_sarif` (both in `src/sarif.rs` and
   in `python/big_code_analysis/_native.pyi`) to document the new
   skip semantics:

   > ``None`` entries (e.g. from per-file ``analyze()`` calls on
   > generated files with the default ``skip_generated=True``)
   > are skipped silently, the same as ``AnalysisError`` entries —
   > both represent files for which no SARIF record should be
   > emitted.

3. Update the `to_sarif` single-input path to also accept `None`?
   Probably not — a bare `bca.to_sarif(None)` is more likely a
   programmer error than the per-file pattern. Keep the
   single-input path raising; only the iterable path filters.

### Tests to Add/Update

In `big-code-analysis-py/tests/test_sarif.py`:

- `test_to_sarif_filters_none_silently` — pass `[good_dict, None,
  another_good_dict]`, assert two findings emitted (one per dict),
  no exception raised. Test-via-revert: removing the `is_none()`
  branch from the loop makes this test fail with the current
  `TypeError`.
- `test_to_sarif_analyze_with_skip_generated_round_trip` — build a
  tempdir with one normal file + one `@generated` file, call
  `[bca.analyze(p) for p in paths]`, pass the resulting
  `[dict, None]` list through `to_sarif`. Assert one SARIF finding
  on the normal file and no error. This is the end-to-end pattern
  the issue body reproduces.
- `test_to_sarif_single_none_still_raises` — pin that
  `bca.to_sarif(None)` (single-input, not iterable) still raises
  `TypeError`. Documents the asymmetry: only the iterable path
  silently skips `None`.

### Verification

- [ ] `cargo test -p big-code-analysis-py` passes
- [ ] `cargo clippy -p big-code-analysis-py` clean
- [ ] `cd big-code-analysis-py && maturin develop && pytest tests/test_sarif.py` passes
- [ ] Manual verification: the reproducer in the issue body succeeds
      with one SARIF finding (was: TypeError)

**Comment by dekobon** [comment]
Fixed in 70718733 on branch `fix/issue-341` (not yet pushed).

**Change:** In `big-code-analysis-py/src/sarif.rs::to_sarif`, the
iterable arm now skips `None` items the same way it skips
`PyAnalysisError`. Both are documented "no record emitted for this
file" outcomes — `AnalysisError` for files we couldn't analyse,
`None` for files classified as generated — so the same silent-skip
semantics apply to both.

**Doc updates:** The Rust pyfunction docstring and the `.pyi` stub
(both the type signature `Iterable[dict | AnalysisError | None]`
and the prose) now explicitly mention `None` tolerance.

**Regression tests** added to `big-code-analysis-py/tests/test_sarif.py`:

- `test_to_sarif_filters_none_entries_silently`: mixed `[dict, None]`
  iterable with `cyclomatic=0` threshold — asserts exactly one
  finding from the dict survives, pinning that `None` is dropped
  *without* taking neighbouring dicts with it.
- `test_to_sarif_does_not_raise_on_pure_none_input`: iterable of
  only `None` yields a well-formed empty SARIF run.

Both tests verified via test-via-revert: against the pre-fix code
they fail with the exact `TypeError: to_sarif expected a result
dict or AnalysisError, got NoneType` from the bug report; against
the fix they pass.

**Validation:** `cargo fmt --check`, `cargo clippy -p
big-code-analysis-py --all-targets -- -D warnings`, `cargo test -p
big-code-analysis-py --all-features` (46 passed), and `pytest
tests/test_sarif.py` (34 passed, including the 2 new ones) all
green.
